util/persistence.py

```diff
@@ -0,0 +1,242 @@
+                key = '%d-%d' % (epoch_index, global_index)
+                if key in checkpoints_lookup:
+                    # we already processed some file(s) for this checkpoint...
+                    _, _, filenames = checkpoints_lookup[key]
+                else:
+                    # creating a new checkpoint tuple
+                    filenames = []
+                    checkpoint = (epoch_index, global_index, filenames)
+                    checkpoints_lookup[key] = checkpoint
+                    checkpoints.append(checkpoint)
+                # add current filename to checkpoint's file-set
+                filenames.append(filename)
+        # return list of checkpoint tuples - ordered by their global index
+        return sorted(checkpoints, key=lambda cp: cp[1])
+
+    def _get_filename(self, filenames):
+        '''
+        Strips the filename-suffix from first file of the passed checkpoint file-set
+        to get a common prefix that can be used to restore the checkpoint.
+        '''
+        return '.'.join(filenames[0].split('.')[:-1])
+
+    def _delete_files(self, filenames):
+        '''
+        Deletes files of a file-set.
+        '''
+        for filename in filenames:
+            log.step('Removing file "%s"...' % filename)
+            os.remove(filename)
+
+    def _prune_and_save(self, session, epoch_index, global_index):
+        '''
+        Removes outdated checkpoint files and stores a new checkpoint.
+        '''
+        keep = self.keep_n_inters if epoch_index == 0 else self.keep_n_epochs
+        assert keep > 0
+        checkpoints = []
+        for checkpoint in self._get_checkpoints():
+            # if checkpoint's global index is greater than this one,
+            if checkpoint[1] > global_index:
+                # it is on an outdated branch of training history and will be deleted
+                log.debug('Removing files of checkpoint "%s", as the global index %d is greater than the new one (%d)...' % \
+                    (self._get_filename(filenames), checkpoint[1], global_index))
+                self._delete_files(filenames)
+            else:
+                # for the next steps we only keep checkpoints of the current
+                # checkpoint family (either epoch ones or intermediate ones)
+                if (epoch_index == 0 and checkpoint[0] == 0) or (epoch_index > 0 and checkpoint[0] > 0):
+                    checkpoints.append(checkpoint)
+        # removing checkpoint files within current checkpoint family that fall off the keep range
+        for _, _, filenames in checkpoints[:-(keep - 1)]:
+            log.debug('Removing files of "%s", as only %d checkpoints should be kept...' % \
+                (self._get_filename(filenames), keep))
+            self._delete_files(filenames)
+        # generating checkpoint filename
+        filename = '%s_%d.ckpt' % (('epoch-%d' % epoch_index) if epoch_index > 0 else 'inter', global_index)
```

Comment:
nit: it would be easier to read with:
```
epoch_index = ('epoch-%d' % epoch_index) if epoch_index > 0 else 'inter'
filename = '%s_%d.ckpt' % (epoch_index, global_index)
```